extensions/amp-story/0.1/amp-story-page.js

```diff
@@ -0,0 +1,642 @@
+/**
+ * Copyright 2017 The AMP HTML Authors. All Rights Reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS-IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+/**
+ * @fileoverview Embeds a story
+ *
+ * Example:
+ * <code>
+ * <amp-story-page>
+ * </amp-story>
+ * </code>
+ */
+import {
+  AnimationManager,
+  hasAnimations,
+} from './animation';
+import {Layout} from '../../../src/layout';
+import {Services} from '../../../src/services';
+import {upgradeBackgroundAudio} from './audio';
+import {dev, user} from '../../../src/log';
+import {EventType, dispatch, dispatchCustom} from './events';
+import {PageElement} from './page-element';
+import {isFiniteNumber} from '../../../src/types';
+import {VideoEvents} from '../../../src/video-interface';
+import {listenOnce} from '../../../src/event-helper';
+
+const LOADING_SCREEN_CONTENTS_TEMPLATE =
+    `<ul class="i-amphtml-story-page-loading-dots">
+      <li class="i-amphtml-story-page-loading-dot"></li>
+      <li class="i-amphtml-story-page-loading-dot"></li>
+      <li class="i-amphtml-story-page-loading-dot"></li>
+    </ul>
+    <p class="i-amphtml-story-page-loading-text">Loading</p>`;
+
+
+/**
+ * CSS class for an amp-story-page that indicates the entire page is loaded.
+ * @const {string}
+ */
+const PAGE_LOADED_CLASS_NAME = 'i-amphtml-story-page-loaded';
+
+
+/**
+ * CSS class for an amp-story-page that indicates the entire page can be shown.
+ * @const {string}
+ */
+const PAGE_SHOWN_CLASS_NAME = 'i-amphtml-story-page-shown';
+
+
+/**
+ * The duration of time (in milliseconds) to show the loading screen for this
+ * page, before showing the page content.
+ * @const {number}
+ */
+const LOAD_TIMEOUT_MS = 8000;
+
+
+/**
+ * The delay (in milliseconds) to wait between polling for loaded resources.
+ */
+const LOAD_TIMER_POLL_DELAY_MS = 250;
```

Comment:
This seems coarse and will probably add ~125ms latency due to imprecision. Is avoiding polling altogether possible?

Reply:
Yeah, should be possible.  The first pass at this implementation used browser events (e.g. [`canplaythrough`](https://developer.mozilla.org/en-US/docs/Web/Events/canplaythrough), [`loadeddata`](https://developer.mozilla.org/en-US/docs/Web/Events/loadeddata), and [`progress`](https://developer.mozilla.org/en-US/docs/Web/Events/progress)) and that didn't work well (seems like in practice, browsers differ pretty greatly in how they implement this, and it results in pretty poor results, especially on Android Chrome).

An intermediate implementation used events from each `PageElement` to the `AmpStoryPage`, but each element polled internally, since there were no events that got us the desired result.  The code written here was a refactor of that to allow only creating a single timer per page, rather than a single timer per element (`n` timers per page).

Reply:
Fair enough. Consider including this under an area for future optimization.

Reply:
Filed #11532